adminpages/memberslist-csv.php

```diff
@@ -143,8 +155,27 @@
 			DISTINCT u.ID
 		FROM $wpdb->users u ";
 
-	if ($s)
-		$sqlQuery .= "LEFT JOIN {$wpdb->usermeta} um ON u.ID = um.user_id ";
+	if ( $s ) {
+		if ( ! empty( $search_key ) ) {
+			// If there's a colon in the search string, make the search smarter.
+			if( in_array( $search_key, array( 'login', 'nicename', 'email', 'url', 'display_name' ) ) ) {
+				$key_column = 'u.user_' . esc_sql( $search_key );
+				$search = " AND $key_column LIKE '%" . esc_sql( $s ) . "%' ";
+			} elseif ( $search_key === 'discount' || $search_key === 'discount_code' || $search_key === 'dc' ) {
+				$user_ids = $wpdb->get_col( "SELECT dcu.user_id FROM $wpdb->pmpro_discount_codes_uses dcu LEFT JOIN $wpdb->pmpro_discount_codes dc ON dcu.code_id = dc.id WHERE dc.code = '" . esc_sql( $s ) . "'" );
+				$search = " AND u.ID IN(" . implode( ",", $user_ids ) . ") ";
```

Comment:
I'm not too concerned but even if we think the value will be an array of numbers, we should probably cast to int anyways like `$user_ids = array_map( 'absint', $user_ids );`

As it is right now, there is no escaping happening on the user IDs before it is inserted into the query and that could become problematic later.

Reply:
I thought about this. In general, I'd say "why not". Might as well be extra careful and escape the var just in case, even if you think it's going to be safe.

In this case, there is a performance hit to absint a potentially large array. There's also little room for the $user_ids var to get out of line. We use it right away on the next line. We should be able to assume the response from $wpdb->get_col() there is false or an array of IDs from the DB.

I'm going to push without it for now and keep this in mind.

Reply:
Another way to put this is I don't see a vector for attack here that doesn't already give you greater access. You'd have to filter the SQL query or alter the DB, and if you could do that, you don't need to mess with this query of ours.